components/PlaybackTask.brs

```diff
@@ -0,0 +1,244 @@
+Library "Roku_Ads.brs" ' Must be initialized in the task (or main) thread
+
+sub init()
+  ? "TRUE[X] >>> PlaybackTask::init()"
+  m.top.functionName = "setup"
+end sub
+
+sub setup() 
+  m.port = createObject("roMessagePort")
+  m.adFacade = m.top.adFacade
+  m.skipAds = false
+  m.isInTruexAd = false
+
+  setupVideo()
+  setupRaf()
+  startPlayback()
+end sub
+
+sub setupVideo()
+  ? "TRUE[X] >>> setupVideo()"
+
+  m.videoPlayer = m.top.video
+
+  videoContent = CreateObject("roSGNode", "ContentNode")
+
+  ' videoContent.url = "http://development.scratch.truex.com.s3.amazonaws.com/roku/simon/roku-reference-app-stream-med.mp4"
+  videoContent.url = "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerFun.mp4"
+  videoContent.title = "video title"
+  videoContent.streamFormat = "mp4"
+  videoContent.length = 600
+
+  videoContent.playStart = 0
+
+  m.videoPlayer.content = videoContent
+  m.videoPlayer.SetFocus(true)
+  m.videoPlayer.visible = true
+  m.videoPlayer.observeFieldScoped("position", m.port)
+  m.videoPlayer.EnableCookies()
+end sub
+
+sub setupRaf()
+  ? "TRUE[X] >>> setupRaf()"
+  raf = Roku_Ads()
+  raf.enableAdMeasurements(true)
+  raf.setContentGenre("Entertainment")
+  raf.setContentId("TrueXSample")
+
+  raf.SetDebugOutput(false) 'debugging
+  raf.SetAdPrefs(false)
+
+  ' adUrl = "pkg:/res/vast.xml"  ' Regular Ads
+  adUrl = "pkg:/res/vast-truex.xml"  ' TrueX preroll + Ads
+  raf.setAdUrl(adUrl)
+  
+  m.adPods = raf.getAds()
+
+  m.raf = raf
+end sub
+
+sub startPlayback()
+  ? "TRUE[X] >>> startPlayback()"
+
+  checkPreroll() ' Currently doesn't handle secondary ad flows (eg. skip choice card).  Requires letting playback to handle
+  if NOT m.isInTruexAd
+    m.videoPlayer.control = "play" 'TODO: Should this go in the flow?
+  end if
+
+  while(true)
+    msg = Wait(0, m.port)
+    msgType = type(msg)
+
+    ' ? "event type:" msgType
+    if msgType = "roSGNodeEvent"
+      field = msg.getField()
+      ? "roSGNodeEvent msg.getField()" field
+      if field = "position" then 
+        onPositionChanged(msg)
```

Comment:
the `wait(0, m.port)` essentially listens for any events that subscribes to the port.  So it is a central way of handling any event based behaviour.  I find it a bit more readable if you have a lot of events and more flexible when building something complex.  As it turns out, I might only care about two events for this because of the scope of the effort, so I wouldn't be against changing the event handlers to be more explicit.

Reply:
Interestingly, using explicit function names for the event callbacks, is run on the render thread.  Since I want to remain in the scope of the task, I'll need to leave it as is for now.